kyuubi-common/src/test/scala/org/apache/kyuubi/session/SessionManagerSuite.scala

```diff
@@ -0,0 +1,155 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.session
+
+import java.time.Duration
+import java.util
+
+import org.apache.hive.service.rpc.thrift._
+import org.apache.thrift.protocol.TBinaryProtocol
+import org.apache.thrift.transport.TSocket
+
+import org.apache.kyuubi.{KyuubiFunSuite, Utils}
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.service.NoopServer
+import org.apache.kyuubi.service.authentication.PlainSASLHelper
+
+class SessionManagerSuite extends KyuubiFunSuite {
+
+  private val server = new NoopServer()
+  private val conf = KyuubiConf()
+    .set(KyuubiConf.FRONTEND_BIND_PORT, 0)
+    .set("kyuubi.test.server.should.fail", "false")
+
+  val user: String = System.getProperty("user.name")
+
+  def startServer(): Unit = {
+    server.initialize(conf)
+    server.start()
+  }
+
+  override def afterAll(): Unit = {
+    server.getServices.foreach(_.stop())
+    super.afterAll()
+  }
+
+  protected def withThriftClient(f: TCLIService.Iface => Unit): Unit = {
+    val hostAndPort = server.connectionUrl.split(":")
+    val host = hostAndPort.head
+    val port = hostAndPort(1).toInt
+    val socket = new TSocket(host, port)
+    val transport =
+      PlainSASLHelper.getPlainTransport(Utils.currentUser, "anonymous", socket)
+
+    val protocol = new TBinaryProtocol(transport)
+    val client = new TCLIService.Client(protocol)
+    transport.open()
+    try {
+      f(client)
+    } finally {
+      socket.close()
+    }
+  }
+
+  protected def withSessionHandle(
+    conf: java.util.Map[String, String]
+  )(f: (TCLIService.Iface, TSessionHandle) => Unit): Unit = {
+    withThriftClient { client =>
+      val req = new TOpenSessionReq()
+      req.setUsername(user)
+      req.setPassword("anonymous")
+      req.setConfiguration(conf)
+      val resp = client.OpenSession(req)
+      val handle = resp.getSessionHandle
+
+      try {
+        f(client, handle)
+      } finally {
+        val tCloseSessionReq = new TCloseSessionReq(handle)
+        try {
+           client.CloseSession(tCloseSessionReq)
+        } catch {
+          case e: Exception => error(s"Failed to close $handle", e)
+        }
+      }
+    }
+  }
+
+  def waitFor(waitForTime: Long)(f: () => Boolean): Unit = {
```

Comment:
can't we simply use `eventually`？

Reply:
I don't know what you mean. Can you give me an example?

Reply:
I mean org.scalatest.concurrent.Eventually

Reply:
Thanks.